### PR TITLE
Add jax.numpy.vectorize

### PR DESCRIPTION
**Pull request by shoyer**
This is basically a non-experimental version of the machinery in `jax.experimental.vectorize`, except:
- It adds the `excluded` argument from NumPy, which works just like `static_argnums` in `jax.jit`.
- It doesn't include the `axis` argument yet (which NumPy doesn't have, and which felt a little half-baked -- it is easy to explicitly use `moveaxis` is desired).

Eventually we might want want to consolidate the specification of signatures with signatures used by shape-checking machinery, but it's nice to emulate NumPy's existing interface, and this is already useful (e.g., for writing vectorized linear algebra routines).

I have added a deprecation warning when the `jax.experimental.vectorize` module
is imported warning that it will go away soon. To make the transition easier, I would vote for waiting for a month or two.

### CODE REVIEW

File: jax/numpy/vectorize.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      dim_sizes = {}
+
+    checked_func = _check_output_dims(
+        excluded_func, dim_sizes, output_core_dims, error_context)
+
+    num_batch_dims = broadcast_args[0].ndim - len(input_core_dims[0])
+    vectorized_func = checked_func
+    for _ in range(num_batch_dims):
+      vectorized_func = api.vmap(vectorized_func)
+    return vectorized_func(*broadcast_args)
```
Comment (shoyer) [inline jax/numpy/vectorize.py]: It occurs to me that there is an alternative implementation here that could potentially be more efficient: instead of broadcasting everything and vmapping everything, we could only broadcast if needed and apply vmap with select batched arguments instead. This could potentially be significantly more efficient in case where partial batching rules exist, e.g., as was implemented in https://github.com/google/jax/pull/2138.

Comment (shoyer) [reply jax/numpy/vectorize.py]: Done.

**Review by mattjj — APPROVED** [review]
AIUI this is an implementation of an [existing `numpy.vectorize` NumPy API](https://numpy.org/doc/1.18/reference/generated/numpy.vectorize.html) (which is always nice because when implementing existing NumPy APIs we don't have to think as hard as for new APIs of our own design), but written in terms of JAX's `vmap` and thus inheriting some nice properties, like not having to call `pyfunc` in a loop (see the docstring @shoyer wrote for `vectorize` for more) and supporting all of JAX's other transformations (like `jit` and `grad`).

In addition to adding support for another part of the NumPy API, which is great for users, this implementation also seems to unlock a more efficient `jnp.linalg.solve` batching rule (also by @shoyer, in a follow-up PR). I can also imagine this serving as a great discussion topic with NumPy developers, and for explaining JAX's `vmap` to more folks in the NumPy/SciPy ecosystem.

I count myself extra lucky for having reviewed this because as an added bonus now I understand the `typing` module a bit better. Thanks, @shoyer !

**Comment by mattjj** [comment]
This also helps us keep `jax.experimental` tidy!

**Comment by shoyer** [comment]
> In addition to adding support for another part of the NumPy API, which is great for users, this implementation also seems to unlock a more efficient `jnp.linalg.solve` batching rule (also by @shoyer, in a follow-up PR)

Actually, I mostly want this for https://github.com/google/jax/pull/2144, which will unlock a more efficient JVP/VJP rule for `jnp.linalg.solve`. (We already have the better batching rule)

**Comment by shoyer** [comment]
thanks @mattjj for the review!
